src/workerd/io/trace.c++

```diff
@@ -596,9 +627,57 @@ void WorkerTracer::log(kj::Date timestamp, LogLevel logLevel, kj::String message
     return;
   }
   trace->bytesUsed = newSize;
+  if (isSpan) {
+    trace->spans.add(timestamp, logLevel, kj::mv(message));
+    trace->numSpans++;
+    return;
+  }
   trace->logs.add(timestamp, logLevel, kj::mv(message));
 }
 
+void WorkerTracer::addSpan(const Span& span, kj::String spanContext) {
+  // This is where we'll actually encode the span for now.
+  // Drop any spans beyond MAX_LIME_SPANS.
+  if (trace->numSpans >= MAX_LIME_SPANS) {
+    return;
+  }
+  if (isPredictableModeForTest()) {
+    // Do not emit span duration information in predictable mode.
+    log(span.endTime, LogLevel::LOG, kj::str("[\"span: ", span.operationName, "\"]"), true);
+  } else {
+    // Time since Unix epoch in seconds, with millisecond precision
+    double epochSecondsStart = (span.startTime - kj::UNIX_EPOCH) / kj::MILLISECONDS / 1000.0;
+    double epochSecondsEnd = (span.endTime - kj::UNIX_EPOCH) / kj::MILLISECONDS / 1000.0;
+    auto message = kj::str("[\"span: ", span.operationName, " ", kj::mv(spanContext), " ",
+        epochSecondsStart, " ", epochSecondsEnd, "\"]");
+    log(span.endTime, LogLevel::LOG, kj::mv(message), true);
+  }
+
+  // TODO(cleanup): Create a function in kj::OneOf to automatically convert to a given type (i.e
+  // String) to avoid having to handle each type explicitly here.
+  for (const Span::TagMap::Entry& tag: span.tags) {
+    auto value = [&]() {
+      KJ_SWITCH_ONEOF(tag.value) {
+        KJ_CASE_ONEOF(str, kj::String) {
+          return kj::str(str);
+        }
+        KJ_CASE_ONEOF(val, int64_t) {
+          return kj::str(val);
+        }
+        KJ_CASE_ONEOF(val, double) {
+          return kj::str(val);
+        }
+        KJ_CASE_ONEOF(val, bool) {
+          return kj::str(val);
+        }
+      }
+      KJ_UNREACHABLE;
+    }();
```

Comment:
Question: should this count towards the max byte limit for the trace?

Reply:
It already does – log() is called to actually write the data and keeps track of `trace->bytesUsed `.